Fix test failures for various combinations of kubectl and Kubernetes versions

skaes opened the pull request:
This patch fixes two different problems:

1. Kubernetes 1.8 has changed the behavior when secrets cannot be found.
   Instead of aborting a deployment quickly, Kubernetes retries quite a
   few times before it gives up. Removed the test case which checks for
   a quick exit.

2. kubectl 1.8 produces slighty different log output for some of the test
   cases. Made the matches version dependent. Another option would be to
   only check the parts which are identical in both versions.

stefanmb commented:
I can confirm this PR fixes all known test failures on k8s `1.8.2` (and also passes on `1.7.6` and `1.7.9`). Test run here: https://buildkite.com/shopify/kubernetes-deploy-gem/builds/206

stefanmb commented:
@skaes Could I ask you to fix the trivial style issues reported here? https://policial.shopify.io/Shopify/kubernetes-deploy/builds/357126

It'll make merging this easier. Thank you!

KnVerey commented:
@skaes any chance you'd be able to update this PR in the next day? We've got our v1.8 CI ready to go now, so we're eager to have this in. 😄

skaes commented on lib/kubernetes-deploy/kubectl.rb:
```diff
@@ -31,5 +31,25 @@ def run(*args, log_failure: nil, use_context: true, use_namespace: true)
+      @version_info ||=
+        begin
+          response, _, status = run("version", "--output=yaml", use_namespace: false, log_failure: true)
+          raise "Could not retrieve kubectl client info" unless status.success?
```
Sure thing. Did not know about KubectlError

skaes commented on lib/kubernetes-deploy/kubectl.rb:
```diff
@@ -31,5 +31,25 @@ def run(*args, log_failure: nil, use_context: true, use_namespace: true)
+    end
+
+    def client_version
+      version_info["clientVersion"]["gitVersion"]
```
I figured that I'll never need an exact comparison. Only comparisons of the form x < 1.8 or x > 1.7. But I can change it, of course.

skaes commented on lib/kubernetes-deploy/kubectl.rb:
```diff
@@ -31,5 +31,25 @@ def run(*args, log_failure: nil, use_context: true, use_namespace: true)
+          response, _, status = run("version", "--output=yaml", use_namespace: false, log_failure: true)
+          raise "Could not retrieve kubectl client info" unless status.success?
+          YAML.load(response)
+        rescue => e
```
Fine with me. But the user might be confused if he sees for example an error from YAML.load.

KnVerey replied:
Ah, I wasn't noticing the YAML.load (even though it's right there 😄 ). I'm fine either way on this one.

skaes commented on lib/kubernetes-deploy/runner.rb:
```diff
@@ -410,12 +410,23 @@ def confirm_context_exists
+      response = nil
       with_retries(2) do
-        _, _, st = kubectl.run("version", use_namespace: false, log_failure: true)
+        response, _, st = kubectl.run("version", "--output=yaml", use_namespace: false, log_failure: true)
```
OK.

skaes commented:
@KnVerey I think I have addressed all your concerns with the latest commits. Would be great to see a new version soon!

BTW, any chance you might merge https://github.com/Shopify/kubernetes-deploy/pull/140 soon? We do have StatefulSets too.

KnVerey commented on lib/kubernetes-deploy/runner.rb:
```diff
@@ -411,10 +411,8 @@ def confirm_context_exists
       with_retries(2) do
-        _, _, st = kubectl.run("version", use_namespace: false, log_failure: true)
-        success = st.success?
+        success = kubectl.version_info rescue false
```
Please rescue specific exceptions (`Psych::SyntaxError` / `KubectlError`)

KnVerey commented:
Actually, [CI](https://buildkite.com/shopify/kubernetes-deploy-gem/builds/216#1bae46e9-4214-46b6-a112-4fc412f7bab4) just finished and we have a problem on 1.6.4:
```

[2017-11-07T21:00:05Z] [WARN][2017-11-07 21:00:05 +0000]	The following command failed: kubectl version --output\=yaml --context\=minikube --request-timeout\=5s
--
  | [2017-11-07T21:00:05Z] [WARN][2017-11-07 21:00:05 +0000]	Error: unknown flag: --output
  | [2017-11-07T21:00:05Z]
  | [2017-11-07T21:00:05Z]
  | [2017-11-07T21:00:05Z] Examples:
  | [2017-11-07T21:00:05Z]   # Print the client and server versions for the current context
  | [2017-11-07T21:00:05Z]   kubectl version
  | [2017-11-07T21:00:05Z]
  | [2017-11-07T21:00:05Z] Options:
  | [2017-11-07T21:00:05Z]   -c, --client=false: Client version only (no server required).
  | [2017-11-07T21:00:05Z]       --short=false: Print just the version number.
  | [2017-11-07T21:00:05Z]
  | [2017-11-07T21:00:05Z] Usage:
  | [2017-11-07T21:00:05Z]   kubectl version [options]
  | [2017-11-07T21:00:05Z]
  | [2017-11-07T21:00:05Z] Use "kubectl options" for a list of global command-line options (applies to all commands).
```

KnVerey commented:
As for the StatefulSet PR, it is almost ready to ship, but we actually need your PR to merge first so that we can exclude rollingUpdate tests on < v1.7. 🙏

skaes commented:
Hopefully this is good to go now.

KnVerey commented:
CI run was green: https://buildkite.com/shopify/kubernetes-deploy-gem/builds/225. Thanks again for the contribution!

stefanmb commented:
@KnVerey That run did not include 1.8, but it was passing previously.

KnVerey commented:
@skaes FYI we released version 0.13.0 yesterday, which includes the StatefulSet support. 😄
